Use an ordered set based on ID for secondary to msm container

lindsayad opened the pull request:
I ran into occasional failures about reusing the base for matrix-free computations on CIVET and in repeated running locally (strangely only when running valgrind (which I should note never errored, it just was the only thing that helped trigger the libMesh failed assertion)). After some bisecting I tracked the change to moving from a vector sequence based on the msm ids to an unordered set based only the msm pointers. Because the msm mesh is rebuilt upon ever residual evaluation the address space used by the msm elements is liable to change and so the hashing and consequently the container iteration order is liable to change. This is what led to the reuse base errors. Because I always expect the number of mortar segments per secondary to be relatively small I think it makes cpu sense to switch to an ordered container and this has the happy side-effect of making the container iteration order deterministic.

Perhaps here is a good place to point out that while the secondary to msm container is an unordered_map, the secondary element pointers do not change from residual to residual and hence the hashing and container iteration order do not change, so that is why we have never seen an issue related to that "unordered-ness" before

Refs #20017

lindsayad committed:
Use an ordered set based on ID for secondary to msm container

I ran into occasional failures about reusing the base for matrix-free
computations on CIVET and in repeated running locally
(strangely only when running valgrind (which I should note never
errored, it just was the only thing that helped trigger the libMesh
failed assertion)). After some bisecting I tracked the change to moving
from a vector sequence based on the msm ids to an unordered set based
only the msm pointers. Because the msm mesh is rebuilt upon ever
residual evaluation the address space used by the msm elements is liable
to change and so the hashing and consequently the container iteration
order is liable to change. This is what led to the reuse base errors.
Because I always expect the number of mortar segments per secondary to
be relatively small I think it makes cpu sense to switch to an ordered
container and this has the happy side-effect of making the container
iteration order deterministic.

Perhaps here is a good place to point out that while the secondary to
msm container is an unordered_map, the secondary element pointers do not
change from residual to residual and hence the hashing and container
iteration order do not change, so that is why we have never seen an
issue related to that "unordered-ness" before

Refs #20017

lindsayad commented:
This will need to wait on #19982 in order to get `CompareDofObjectsByID`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/936765/) on ba09ba4 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/20084/site)

This comment will be updated on new commits.

lindsayad commented:
Problematic test: `contact/test:bouncing-block-contact/ping-ponging.mortar-no-ping-pong_weighted`

moosebuild commented:
Job [Test Heavy](https://civet.inl.gov/job/936774/) on ba09ba4 : invalidated by @lindsayad

Let&#39;&#39;s see if this is a sporadic failure ... which makes it no better

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/936775/) on ba09ba4 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>e9b275</th>
      <th colspan=4>#20084 ba09ba</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.31%</td>
      <td>82.31%</td>
      <td>-0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>73298</td>
      <td>73296</td>
      <td>-2</td>
      <td>2</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15758</td>
      <td>15758</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20084/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20084/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/20084/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/20084/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/20084/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/20084/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/20084/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/20084/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/20084/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/20084/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/20084/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/20084/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/20084/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/20084/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/20084/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/20084/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/20084/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/20084/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/20084/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/20084/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/20084/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/20084/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/20084/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/20084/coverage/tensor_mechanics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/20084/coverage/xfem)

</details>



This comment will be updated on new commits.

recuero commented:
Basically, never use pointers to keep track of entities in the mortar segment mesh? Or am I oversimplifying...

Maybe that failing test was on the limit? Wonder if a slightly smaller time step would help there.

dschwen commented:
Pointers at bad keys as their order can vary from run to run and machine to machine.

lindsayad commented:
Just because an algorithm produces a slightly different answer from run-to-run or even from residual-to-residual does not mean, in my opinion, that it is a bad algorithm. If the differences are due to things like different iteration order through an `unordered_set` or different order through a `set` that uses pointers as keys, then there is nothing innately wrong with having differences. To me you are generating results in a distribution with no result really being more right or wrong than the other. You can choose to use a `set` with ids as keys and then get the same answer every time ... but that just means you are hitting one part of a statistical distribution and the result is no more right (even though you get it every time) than if you used an `unordered_set` or a `set` with pointer keys.

Now perhaps you can argue that getting the same result every time is a nice thing from a user perspective, and that is a fair argument. But even then the best you can hope for is getting the same answer every time ***in serial***. Once you add multiple processes determinism goes out the window again. That is exactly why we only run the "reuse base" check in serial ... and as the recent mortar changes showed, even in serial you can still have a "good" function evaluation that is not deterministic.

So I think my statement
> Because I always expect the number of mortar segments per secondary to be relatively small I think it makes cpu sense to switch to an ordered container and this has the happy side-effect of making the container iteration order deterministic.

is about as well as I can put it ... it makes algorithmic sense to use a tree container because in many cases there will only be sets of one mortar element per secondary element and at most probably never much more than 10, in which case I think it makes sense to use a tree instead of a hash table ... and if I"m going back to `set` than I might as well sort by ids and have the determinism because it has the very nice side-effect that it won't trigger the reuse base error. I thought about removing the reuse base error out of libmesh but as @fdkong has pointed out and as [today's slack thread](https://moosedevelopers.slack.com/archives/C01054VRUEM/p1642622594121300) shows, most of the time the error check catches very real errors.

lindsayad commented:
> Maybe that failing test was on the limit? Wonder if a slightly smaller time step would help there.

I checked the condition number for the tests that had been failing on next and they were quite good ... no more than 1e3. It's just that the vector equivalence check in libMesh is very sensitive. I think it might make some sense, if we can, to loosen the vector comparison tolerances such that things like using unordered containers don't trigger errors.

lindsayad committed:
Use -snes_ksp_ew option with ping pong weighted test

I was mortified when I discovered that somehow changing from an
unordered to ordered container magically doubled the nonlinear iteration
count for one timestep and created warnings where none had been before.
I spent all afternoon trying to figure out what malicious bug I had let
in in such a small commit. And ultimately I think I've realized that
*this* test is a test that's right on the edge. By adding `-snes_ksp_ew`
to this test it actually runs *better* than the previous container type
(well at least in comparison to the old vector ... didn't check against
the unordered_set). And in general with `-snes_ksp_ew` this test runs
better for all container types I exammined than if those container types
were run without `-snes_ksp_ew`. So I think `-snes_ksp_ew` in general is
an excellent choice for this particular test (and perhaps at least
frictionless mortar contact in general ... but I would want to run it on
a lot more inputs before I stated that definitively)

lindsayad commented:
@loganharbour this should take care of 2/3 of the next failures. I will look into the perf graph valgrind failure tomorrow

moosebuild commented:
Job [Moose Install Target Tests](https://civet.inl.gov/job/937357/) on 2b3f139 : invalidated by @lindsayad
